src/quote/leveraged-extended/provider.ts

```diff
@@ -0,0 +1,221 @@
+}
+
+export class LeveragedExtendedQuoteProvider
+  implements
+    QuoteProvider<
+      FlashMintLeveragedExtendedQuoteRequest,
+      FlashMintLeveragedExtendedQuote
+    >
+{
+  constructor(
+    private readonly provider: JsonRpcProvider,
+    private readonly zeroExApi: ZeroExApi
+  ) {}
+
+  async getQuote(
+    request: FlashMintLeveragedExtendedQuoteRequest
+  ): Promise<FlashMintLeveragedExtendedQuote | null> {
+    const { provider, zeroExApi } = this
+    const { inputToken, indexTokenAmount, isMinting, outputToken, slippage } =
+      request
+    const indexToken = isMinting ? outputToken : inputToken
+    const indexTokenSymbol = indexToken.symbol
+    const includedSources = getIncludedSources()
+    const network = await provider.getNetwork()
+    const chainId = network.chainId
+    console.log('chainId:', chainId)
+
+    const leveragedTokenData = await getLeveragedTokenData(
+      indexToken.address,
+      indexTokenAmount,
+      indexTokenSymbol,
+      isMinting,
+      chainId,
+      provider
+    )
+    console.log(leveragedTokenData)
+    if (leveragedTokenData === null) return null
+
+    const debtCollateralResult = isMinting
+      ? await getSwapDataDebtCollateral(
+          leveragedTokenData,
+          includedSources,
+          slippage,
+          chainId,
+          zeroExApi
+        )
+      : await getSwapDataCollateralDebt(
+          leveragedTokenData,
+          includedSources,
+          slippage,
+          chainId,
+          zeroExApi
+        )
+    if (!debtCollateralResult) return null
+    const { collateralObtainedOrSold } = debtCollateralResult
+    const { swapDataDebtCollateral } = debtCollateralResult
+    // Relevant when issuing
+    const collateralShortfall = leveragedTokenData.collateralAmount.sub(
+      collateralObtainedOrSold
+    )
+    // Relevant when redeeming
+    const leftoverCollateral = leveragedTokenData.collateralAmount.sub(
+      collateralObtainedOrSold
+    )
+    const inputOutputTokenAddress = getPaymentTokenAddress(
+      isMinting ? inputToken.address : outputToken.address,
+      isMinting ? inputToken.symbol : outputToken.symbol
+    )
+    const { swapDataPaymentToken, paymentTokenAmount } =
+      await getSwapDataAndPaymentTokenAmount(
+        indexTokenSymbol,
+        leveragedTokenData.collateralToken,
+        collateralShortfall,
+        leftoverCollateral,
+        inputOutputTokenAddress,
+        isMinting,
+        slippage,
+        includedSources,
+        zeroExApi,
+        chainId
+      )
+
+    const estimatedInputOutputAmount = paymentTokenAmount
+    const inputOuputTokenDecimals = isMinting
+      ? inputToken.decimals
+      : outputToken.decimals
+    const inputOutputTokenAmount = slippageAdjustedTokenAmount(
+      estimatedInputOutputAmount,
+      inputOuputTokenDecimals,
+      slippage,
+      isMinting
+    )
+    return {
+      inputTokenAmount: isMinting ? inputOutputTokenAmount : indexTokenAmount,
```

Comment:
Can be simplified as soon as `inputTokenAmount` is part of the quote request type.